utils/swlite-initialize.1
.\" swlite-initialize - create an admin authorization and extension in sqlite backend
.\" Copyright (C) 2017-2018 Tycho Softworks
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH swlite-initialize "1" "December 2017" "SipWitchQt" "Tycho Softworks"
.SH NAME
swlite-initialize \- create an admin authorization and extension in sqlite backend
.SH SYNOPSIS
.B swlite-initialize
.SH DESCRIPTION
The purpose of swlite-initialize is to take a newly created and empty sqlite
database, such as the one created by default by the server when installed from
a package, and add a user authorization that has full administration access.
Using this initial user all other admin tasks, including creating additional
users, can then be performed from the desktop client.

If you are using mysql then you would use the ipl-sipwitch utility instead.
.SH OPTIONS
.TP
.B \-\-sha256
Use a SHA-256 digest when setting a password (default is MD5).
.TP
.B \-\-sha512
Use a SHA-512 digest when setting a password (default is MD5).
.SH AUTHOR
.B swlite-initialize
was written by David Sugar <tychosoft@gmail.com>.
.SH "REPORTING BUGS"
Report bugs to tychosoft@gmail.com.
.SH COPYRIGHT
Copyright \(co 2017 Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



CHANGELOG
v0.3.0
* Improved unit file
* Added security model checks; no extension keys implemented yet
* Admin and suspend states and modifications supported
* Added admin drop/disconnect
* Various ui improvements

v0.2.7
* Improved packaging support
* Updated copyright notices
* Utils use stand-alone to init mysql servers
* De-emphasizing sqlite beyond testing builds
* Device Management and verification support added 

v0.2.6
* Simplified expiration handling
* Re-arranged font selection
* Introduced desktop & audio notify setting
* Integrated login auto start behavior
* Speed dial ui support
* Call coverage setting support
* Call forwarding setting support
* Delay ringing option supported
* Per user and global auto answer settings
* Start of integrated voice mail
* Device keys introduced

v0.2.5
* Customization at build time
* Topic management in sessions
* Policy setting for main public room

v0.2.4
* Session/state based toolbar management
* Online status of all extensions shown
* Membership management for groups
* Server system notification messaging
* Delete authorizations supported

v0.2.3
* Schema changes
* Message expiration

v0.2.2
* Change password support
* Simple markdown support
* Zeroconfg support
* Message sync for existing extensions
* Functional testing support
* Improved adduser ui
* Expiration marking on client

v0.2.1
* Fix for crash sending offline messages
* Visual styling improvements and optimizations

V0.2.0
* IPL loading of tables, such as for external mysql db
* User classification of priviledges
* Ability to edit extension names from profile
* Clickable and highlighted urls in messages
* Offline message sync
* Cleaner db handling for login/logout and multiple dbs
* Client import/export
* Font picker for message view
* Admin can create users from client
* Phonebook expanded
* New message indicators

v0.1.4
* I want to send a message to someone
* I Want to receive messages
* I want to search my messages
* I want to logout, and back in
* I want to know what extensions I can login as
* I want the ui to become available in language of my choice
* I want to know what devices are on my extension
* I want to autostart when I login on linux
* sipwitchqt message laying and rendering engine introduced
* email and public uri properties added to contacts

v0.1.3
* Roster processing on client and server
* Search processing on sessions screen
* click feedback, doubleclick contacts
* vibrant and light icon schemes

v0.1.2
* Start of sessions page and status

v0.1.1
* udp for eXosip2 messaging support
* automatic authentication integrated
* Start of core messaging code
* Replacing json with xdp
* Start of contacts (phonebook) page

v0.1.0
* Core authentication complete 
* Core registration with expiration
* Extended registration support for sipwitchqt clients

v0.0.2
* More complete test harness debug config
* Reset and dialogs for client
* Release packaging for primary platforms

v0.0.1
* Initial development testing release



utils/swlite-initialize.rb
#!/usr/bin/env ruby
# Copyright (C) 2017-2018 Tycho Softworks
#
# This file is free software; as a special exception the author gives
# unlimited permission to copy and/or distribute it, with or without
# modifications, as long as this notice is preserved.

['optparse', 'io/console', 'sqlite3', 'digest'].each {|mod| require mod}

RESERVED_NAMES = ['operators', 'system', 'anonymous', 'nobody']
database = 'sqlite'
digest_type = 'MD5'
first = '100'
last = '699'

if STDIN.respond_to?(:noecho)
  def get_pass(prompt="Password: ")
    print prompt
    input = STDIN.noecho(&:gets).chomp
    exit unless input.size > 0
    print "\n"
    input
  end
else
  def get_pass(prompt="Password: ")
    input = `read -s -p "#{prompt}" password; echo $password`.chomp
    exit unless input.size > 0
    print "\n"
    input
  end
end

def get_input(*args)
  print(*args)
  input = gets.chomp
  exit unless input.size > 0
  input
end

def get_secret(type, user, domain, pass)
  case type
  when 'SHA-256'
    Digest::SHA256.hexdigest "#{user}:#{domain}:#{pass}"
  when 'SHA-512'
    Digest::SHA512.hexdigest "#{user}:#{domain}:#{pass}"
  else
    Digest::MD5.hexdigest "#{user}:#{domain}:#{pass}"
  end
end

domain = nil
dbname = '/var/lib/sipwitchqt/local.db'
dbname = '../testdata/local.db' if File.writable?('../testdata/local.db')
dbname = '../userdata/local.db' if File.writable?('../userdata/local.db')
abort("*** swlite-initialize: no database") unless File.writable?(dbname)

OptionParser.new do |opts|
  opts.banner = 'Usage: swlite-initialize'

  opts.on('-2', '--sha256', 'use sha-256') do
    digest_type = 'SHA-256'
  end

  opts.on('-5', '--sha512', 'use sha-512') do
    digest_type = 'SHA-512'
  end
end.parse!
abort(opts.banner) if(ARGV.size > 0)

begin 
  db = SQLite3::Database.open dbname
  db.execute("SELECT realm, dialplan FROM Config") do |row|
    abort("*** swlite-initialize: multiple config entries") unless domain == nil
    domain = row[0]
    dialing = row[1]
    case dialing
    when 'STD3'
      first = '100'
      last = '699'
    else
      abort("*** swlite-initialize: #{dialing}: invalid dialing plan used")
    end
  end
rescue SQLite3::SQLException
  abort("*** swlite-initialize: no switch table")
end

print "Realm #{domain}\n" if domain != nil
realm = domain if domain != nil

begin
  realm   = get_input "Server realm: " if realm === nil
  user    = get_input "Authorizing user: "
  abort("*** ipl-sipwitch: #{user}: reserved name") if RESERVED_NAMES.include?(user)
  extnbr  = get_input "User extension:   "
  abort("*** ipl-sipwitch: ext must be #{first}-#{last}") unless extnbr >= first and extnbr <= last
  display = get_input "Display name:     "
  email = get_input   "Email address:    "
rescue Interrupt
  abort("")
  exit
end

begin
  print "\nEnter a password for #{extnbr} to authorize with\n"
  pass1    = get_pass "Password: "
  pass2    = get_pass "Verify:   "
  abort("*** ipl-sipwitch: passwords do not match") unless pass1 == pass2
  extpass = get_secret(digest_type,user,realm,pass1)
rescue Interrupt
  abort("^C")
  exit
end

begin
  db = SQLite3::Database.open dbname
  db.execute("INSERT INTO Authorize(authname,authdigest,realm,secret,fullname,authtype,authaccess,email) VALUES('#{user}','#{digest_type}','#{realm}','#{extpass}','#{display}','USER','REMOTE','#{email}')")
  db.execute("INSERT INTO Extensions(extnbr,authname) VALUES(#{extnbr},'#{user}')")
  db.execute("INSERT INTO Endpoints(extnbr) VALUES(#{extnbr})")
  db.execute("INSERT INTO Admin(authname,extnbr) VALUES('system',#{extnbr})")
rescue SQLite3::SQLException
  abort("*** swlite-initialize: cannot initialize #{extnbr}")
end

